query-parser/src/parse.rs

```diff
@@ -83,320 +96,348 @@ error_chain! {
+    choice::<[&mut Parser<Input = _, Output = FindSpec>; 4], _>
         ([&mut try(Find::find_scalar()),
           &mut try(Find::find_coll()),
           &mut try(Find::find_tuple()),
           &mut try(Find::find_rel())])
-        .parse_stream(input)
 });
 
-// Parse a sequence of values into one of four find specs.
-//
-// `:find` must be an array of plain var symbols (?foo), pull expressions, and aggregates.
-// For now we only support variables and the annotations necessary to declare which
-// flavor of :find we want:
-//
-//
-//     `?x ?y ?z  `     = FindRel
-//     `[?x ...]  `     = FindColl
-//     `?x .      `     = FindScalar
-//     `[?x ?y ?z]`     = FindTuple
-//
-#[allow(dead_code)]
-pub fn find_seq_to_find_spec(find: &[edn::Value]) -> FindParseResult {
-    Find::find()
-        .parse(find)
-        .map(|x| x.0)
-        .map_err::<ValueParseError, _>(|e| e.translate_position(find).into())
-        .map_err(|e| Error::from_kind(ErrorKind::FindParseError(e)))
+/// TODO: extract macro for matching these `Keyword` instances.
+def_parser!(Find, literal_find, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "find"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_with, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "with"
+        } else {
+            false
+        }
+    })
+});
+
+def_parser!(Find, literal_where, edn::ValueAndSpan, {
+    satisfy(|v: edn::ValueAndSpan| {
+        if let edn::SpannedValue::Keyword(ref s) = v.inner {
+            s.0.as_str() == "where"
+        } else {
+            false
+        }
+    })
+});
+
+/// Express something close to a builder pattern for a `FindQuery`.
+enum FindQueryPart {
+    FindSpec(FindSpec),
+    With(Vec<Variable>),
+    WhereClauses(Vec<WhereClause>),
 }
 
-#[allow(dead_code)]
-pub fn clause_seq_to_patterns(clauses: &[edn::Value]) -> WhereParseResult {
-    Where::clauses()
-        .parse(clauses)
+/// This is awkward, but will do for now.  We use `keyword_map()` to optionally accept vector find
+/// queries, then we use `FindQueryPart` to collect parts that have heterogeneous types; and then we
+/// construct a `FindQuery` from them.
+def_parser!(Find, query, FindQuery, {
+    let p_find_spec = Find::literal_find()
+        .with(vector().of(Find::spec().map(FindQueryPart::FindSpec)));
+
+    let p_with_vars = Find::literal_with()
+        .with(vector().of(many(Query::variable()).map(FindQueryPart::With)));
+
+    let p_where_clauses = Find::literal_where()
+        .with(vector().of(Where::clauses().map(FindQueryPart::WhereClauses))).expected(":where clauses");
+
+    (or(map(), keyword_map()))
+        .of(many(choice::<[&mut Parser<Input = ValueStream, Output = FindQueryPart>; 3], _>([
```

Comment:
I'm going to punt this to follow-up.  I've lost the will to live with this patch :(

https://github.com/mozilla/mentat/issues/401